ibm_kms_instance_policies fails when trying to communicate with a private only enabled Key Protect instance

There seems to be a bug with the [ibm_kms_instance_policies](https://registry.terraform.io/providers/IBM-Cloud/ibm/latest/docs/resources/kms_instance_policies) provider resource when it tries to run on a “private-only” Key Protect instance. The error returned is:
```
Error: [ERROR] Get Policies failed with error : kp.Error: correlation_id='964cfd54-cfca-4b64-bfc5-f79c23a1806a', msg='Unauthorized: Either the user does not have access to the specified resource, the resource does not exist, or the region is incorrectly set
```
I am running in a Schematics environment on a VRF enabled account, so it should have access to the private endpoint. When I set the Key Protect instance to allow “public-and-private”, the error does not occur.

Other resources (such as [ibm_kms_key](https://registry.terraform.io/providers/IBM-Cloud/ibm/latest/docs/resources/kms_key)) have an `endpoint_type` argument that can be set to public or private to allow you to choose which endpoint to use. But no such argument exists for `ibm_kms_instance_policies` so I’m wondering by any chance is the resource hard coded to use public endpoint, and hence failing here?

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

<!--- Thank you for keeping this note for the community --->

### Terraform CLI and Terraform IBM Provider Version

tf 1.5.7
ibm provider 1.62.0

### Affected Resource(s)

<!--- Please list the affected resources and data sources. --->

* ibm_kms_instance_policies

### Terraform Configuration Files

The code is in root level of code in https://github.com/terraform-ibm-modules/terraform-ibm-base-security-services-da/pull/2

### Debug Output

[test.log.zip](https://github.com/IBM-Cloud/terraform-provider-ibm/files/14402783/test.log.zip)


### Panic Output

<!--- If Terraform produced a panic, please provide a link to a GitHub Gist containing the output of the `crash.log`. --->

### Expected Behavior

No error

### Actual Behavior

```
Error: [ERROR] Get Policies failed with error : kp.Error: correlation_id='964cfd54-cfca-4b64-bfc5-f79c23a1806a', msg='Unauthorized: Either the user does not have access to the specified resource, the resource does not exist, or the region is incorrectly set
```

### Steps to Reproduce

<!--- Please list the steps required to reproduce the issue. --->

1. `terraform apply`

### Important Factoids

(See description at top of this issue)

### References

<!---
Information about referencing Github Issues: https://help.github.com/articles/basic-writing-and-formatting-syntax/#referencing-issues-and-pull-requests

Are there any other GitHub issues (open or closed) or pull requests that should be linked here? Vendor documentation? For example:
--->

* #0000

After updating to use allowed network flag, I am able to recreate the error
```
Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

ibm_resource_instance.kp_instance: Creating...
ibm_resource_instance.kp_instance: Still creating... [10s elapsed]
ibm_resource_instance.kp_instance: Creation complete after 13s [id=crn:v1:bluemix:public:kms:us-south:a/eba0f7b1166e441ab74ac94e564c72ec:fa971212-e0c7-4fcb-87c6-82e3d5f1bd46::]
ibm_kms_key_rings.key_ring: Creating...
ibm_kms_instance_policies.instance_policy: Creating...
╷
│ Error: [ERROR] Error while creating key ring : kp.Error: correlation_id='98283df4-c85c-4035-b693-1708194ee3b8', msg='Unauthorized: Either the user does not have access to the specified resource, the resource does not exist, or the region is incorrectly set'
│ 
│   with ibm_kms_key_rings.key_ring,
│   on main.tf line 10, in resource "ibm_kms_key_rings" "key_ring":
│   10: resource "ibm_kms_key_rings" "key_ring" {
│ 
╵
╷
│ Error: [ERROR] Get Policies failed with error : kp.Error: correlation_id='86ef5e93-2af1-4e41-bc85-285f71404859', msg='Unauthorized: Either the user does not have access to the specified resource, the resource does not exist, or the region is incorrectly set'
│ 
│   with ibm_kms_instance_policies.instance_policy,
│   on main.tf line 25, in resource "ibm_kms_instance_policies" "instance_policy":
│   25: resource "ibm_kms_instance_policies" "instance_policy" {
│ 
╵
wsiew@cloudshell:~$ 
```

--- a/ibm/service/kms/resource_ibm_kms_instance_policies.go
+++ b/ibm/service/kms/resource_ibm_kms_instance_policies.go
@@ -36,6 +36,13 @@ func ResourceIBMKmsInstancePolicy() *schema.Resource {
 				Description:      "Key protect or hpcs instance GUID or CRN",
 				DiffSuppressFunc: suppressKMSInstanceIDDiff,
 			},
+			"endpoint_type": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				Computed:     true,
+				ValidateFunc: validate.ValidateAllowedStringValues([]string{"public", "private"}),
+				Description:  "public or private",
+			},
 			"dual_auth_delete": {
 				Type:         schema.TypeList,
 				Optional:     true,

--- a/ibm/service/kms/resource_ibm_kms_key.go
+++ b/ibm/service/kms/resource_ibm_kms_key.go
@@ -6,6 +6,7 @@ package kms
 import (
 	"context"
 	"fmt"
+	"log"
 	"net/url"
 	"strconv"
 	"strings"
@@ -139,6 +140,31 @@ func ResourceIBMKmskey() *schema.Resource {
 				Computed:    true,
 				Description: "Key protect or hpcs instance CRN",
 			},
+
+			"registrations": {
+				Type:        schema.TypeList,
+				Computed:    true,
+				Description: "Registrations of the key across different services",
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"key_id": {
+							Type:        schema.TypeString,
+							Computed:    true,
+							Description: "The key id being used in the policy",
+						},
+						"resource_crn": {
+							Type:        schema.TypeString,
+							Computed:    true,
+							Description: "The resource crn tied to the registration of the key",
+						},
+						"prevent_key_deletion": {
+							Type:        schema.TypeBool,
+							Computed:    true,
+							Description: "Determines if the registration of a key prevents a deletion.",
+						},
+					},
+				},
+			},
 			flex.ResourceName: {
 				Type:        schema.TypeString,
 				Computed:    true,
@@ -225,6 +251,10 @@ func resourceIBMKmsKeyDelete(d *schema.ResourceData, meta interface{}) error {
 
 	_, err1 := kpAPI.DeleteKey(context.Background(), keyid, kp.ReturnRepresentation, f)
 	if err1 != nil {
+		registrations := d.Get("registration").([]map[string]interface{})
+		if registrations != nil && len(registrations) > 0 {
+			log.Fatalf("The key_id %s has the following registrations: %v\n", keyid, registrations)
+		}
 		return fmt.Errorf("[ERROR] Error while deleting: %s", err1)
 	}
 	d.SetId("")
@@ -327,6 +357,23 @@ func setKeyDetails(d *schema.ResourceData, meta interface{}, instanceID string, 
 
 	d.Set(flex.ResourceControllerURL, rcontroller+"/services/kms/"+url.QueryEscape(crn1)+"%3A%3A")
 
+	// Get the Registration of the key
+	registrations, err := kpAPI.ListRegistrations(context.Background(), key.ID, "")
+	if err != nil {
+		return err
+	}
+	// making a map[string]interface{} for terraform key.registration Attribute
+	rSlice := make([]map[string]interface{}, 0)
+	for _, r := range registrations.Registrations {
+		registration := map[string]interface{}{
+			"key_id":               r.KeyID,
+			"resource_crn":         r.ResourceCrn,
+			"prevent_key_deletion": r.PreventKeyDeletion,
+		}
+		rSlice = append(rSlice, registration)
+	}
+	d.Set("registrations", rSlice)
+
 	return nil
 }
 
@@ -396,7 +443,8 @@ func populateSchemaData(d *schema.ResourceData, meta interface{}) (*kp.Client, e
 		return nil, err
 	}
 	// keyid := d.Id()
-	key, err := kpAPI.GetKey(context.Background(), keyid)
+	ctx := context.Background()
+	key, err := kpAPI.GetKey(ctx, keyid)
 	if err != nil {
 		kpError := err.(*kp.Error)
 		if kpError.StatusCode == 404 || kpError.StatusCode == 409 {
@@ -413,5 +461,6 @@ func populateSchemaData(d *schema.ResourceData, meta interface{}) (*kp.Client, e
 	if err != nil {
 		return nil, err
 	}
+
 	return kpAPI, nil
 }

--- a/ibm/service/kms/resource_ibm_kms_key.go
+++ b/ibm/service/kms/resource_ibm_kms_key.go
@@ -6,7 +6,6 @@ package kms
 import (
 	"context"
 	"fmt"
-	"log"
 	"net/url"
 	"strconv"
 	"strings"
@@ -251,11 +250,17 @@ func resourceIBMKmsKeyDelete(d *schema.ResourceData, meta interface{}) error {
 
 	_, err1 := kpAPI.DeleteKey(context.Background(), keyid, kp.ReturnRepresentation, f)
 	if err1 != nil {
-		registrations := d.Get("registration").([]map[string]interface{})
+		registrations := d.Get("registrations").([]interface{})
+		var registrationLog error
 		if registrations != nil && len(registrations) > 0 {
-			log.Fatalf("The key_id %s has the following registrations: %v\n", keyid, registrations)
+			resourceCrns := make([]string, 0)
+			for _, registration := range registrations {
+				r := registration.(map[string]interface{})
+				resourceCrns = append(resourceCrns, r["resource_crn"].(string))
+			}
+			registrationLog = fmt.Errorf(". The key has the following active registrations which may interfere with deletion: %v", resourceCrns)
 		}
-		return fmt.Errorf("[ERROR] Error while deleting: %s", err1)
+		return fmt.Errorf("[ERROR] Error while deleting: %s%s", err1, registrationLog)
 	}
 	d.SetId("")
 	return nil

--- a/ibm/service/kms/resource_ibm_kms_key.go
+++ b/ibm/service/kms/resource_ibm_kms_key.go
@@ -149,17 +149,17 @@ func ResourceIBMKmskey() *schema.Resource {
 						"key_id": {
 							Type:        schema.TypeString,
 							Computed:    true,
-							Description: "The key id being used in the policy",
+							Description: "The id of the key being used in the registration",
 						},
 						"resource_crn": {
 							Type:        schema.TypeString,
 							Computed:    true,
-							Description: "The resource crn tied to the registration of the key",
+							Description: "The CRN of the resource tied to the key registration",
 						},
 						"prevent_key_deletion": {
 							Type:        schema.TypeBool,
 							Computed:    true,
-							Description: "Determines if the registration of a key prevents a deletion.",
+							Description: "Determines if the registration of the key prevents a deletion.",
 						},
 					},
 				},

--- a/website/docs/r/kms_instance_policies.html.markdown
+++ b/website/docs/r/kms_instance_policies.html.markdown
@@ -57,6 +57,8 @@ The following arguments are supported:
 
 
 - `instance_id` - (Required, String) The key-protect instance ID for creating policies.
+- `endpoint_type` - (Optional, String) The type of the public endpoint, or private endpoint to be used for creating keys.
+
 - `rotation` - (Optional,list) The Instance rotation time interval in months, with a minimum of 1, and a maximum of 12.
   Nested scheme for `rotation`:
 

--- a/website/docs/r/kms_key.html.markdown
+++ b/website/docs/r/kms_key.html.markdown
@@ -131,6 +131,13 @@ In addition to all argument reference list, you can access the following attribu
 - `key_id` - (String) The ID of the key.
 - `key_ring_id` - (String) The ID of the key ring that your Key Protect key belongs to.
 - `type` - (String) The type of the key KMS or HPCS.
+- `registrations` - (List) The registrations associated with the key.
+
+  Nested scheme for `registrations`:
+  - `key_id` - (String) The id of the key associated with the association.
+  - `resource_crn` - (String) The CRN of the resource that has a registration to the key.
+  - `prevent_key_deletion` - (Boolean) Determines if the resource prevents the key deletion.
+
 - `policy` - (String) The policies associated with the key.
 
   Nested scheme for `policy`:
@@ -154,6 +161,7 @@ In addition to all argument reference list, you can access the following attribu
      - `id` - (String) The v4 UUID used to uniquely identify the policy resource, as specified by RFC 4122.
      - `last_update_date` - (Timestamp)  The date when the policy last replaced or modified. The date format follows RFC 3339.
      - `updated_by` - (String) The unique ID for the resource that updated the policy.
+
 
 ## Import
 The `ibm_kms_key` can be imported by using the `id` and `crn`.

